AssertionError in is_package_downloaded

```julia
(AbstractPlotting) pkg> up
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `git@github.com:JuliaRegistries/General.git`
    Updating registry at `~/.julia/registries/HolyLabRegistry`
    Updating git-repo `git@github.com:HolyLab/HolyLabRegistry.git`
No Changes to `~/.julia/dev/AbstractPlotting/Project.toml`
No Changes to `~/.julia/dev/AbstractPlotting/Manifest.toml`
ERROR: AssertionError: sourcepath !== nothing
Stacktrace:
  [1] is_package_downloaded(ctx::Pkg.Types.Context, pkg::Pkg.Types.PackageSpec)
    @ Pkg.Operations ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:1757
  [2] #12
    @ ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:118 [inlined]
  [3] _all
    @ ./reduce.jl:923 [inlined]
  [4] #all#696
    @ ./reducedim.jl:886 [inlined]
  [5] all
    @ ./reducedim.jl:886 [inlined]
  [6] is_instantiated(ctx::Pkg.Types.Context)
    @ Pkg.Operations ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:118
  [7] instantiate(ctx::Pkg.Types.Context; manifest::Nothing, update_registry::Bool, verbose::Bool, platform::Base.BinaryPlatforms.Platform, allow_autoprecomp::Bool, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.API ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:1301
  [8] precompile(ctx::Pkg.Types.Context; internal_call::Bool, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.API ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:917
  [9] _auto_precompile(ctx::Pkg.Types.Context)
    @ Pkg.API ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:902
 [10] #up#38
    @ ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:74 [inlined]
 [11] up(pkgs::Vector{Pkg.Types.PackageSpec})
    @ Pkg.API ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:72
 [12] do_cmd!(command::Pkg.REPLMode.Command, repl::REPL.LineEditREPL)
    @ Pkg.REPLMode ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/REPLMode/REPLMode.jl:408
 [13] do_cmd(repl::REPL.LineEditREPL, input::String; do_rethrow::Bool)
    @ Pkg.REPLMode ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/REPLMode/REPLMode.jl:386
 [14] do_cmd
    @ ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/REPLMode/REPLMode.jl:377 [inlined]
 [15] (::Pkg.REPLMode.var"#24#27"{REPL.LineEditREPL, REPL.LineEdit.Prompt})(s::REPL.LineEdit.MIState, buf::IOBuffer, ok::Bool)
    @ Pkg.REPLMode ~/src/julia-master/usr/share/julia/stdlib/v1.6/Pkg/src/REPLMode/REPLMode.jl:550
 [16] #invokelatest#2
    @ ./essentials.jl:707 [inlined]
 [17] invokelatest
    @ ./essentials.jl:706 [inlined]
 [18] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
    @ REPL.LineEdit ~/src/julia-master/usr/share/julia/stdlib/v1.6/REPL/src/LineEdit.jl:2441
 [19] run_frontend(repl::REPL.LineEditREPL, backend::REPL.REPLBackendRef)
    @ REPL ~/src/julia-master/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:1126
 [20] (::REPL.var"#44#49"{REPL.LineEditREPL, REPL.REPLBackendRef})()
    @ REPL ./task.jl:406

```

Project and Manifest in [this gist](https://gist.github.com/timholy/03e2df389715b95a27fefbaac287c31e). (But I have lots of locally devved packages with changes, so this may be hard to replicate.)

Thanks. Pkg fails to lookup the source of `MosaicViews` since it is not in the manifest. Compare e.g.
```
┌ Info: is_package_downloaded
│   project_file = "/home/fredrik/dev/pkg-dbg/MosaicViews/Project.toml"
│   pkg =
│    PackageSpec(
│      name = PaddedViews
│      uuid = 5432bcbf-9aad-5242-b902-cca2824c8663
│      tree_hash = 91d229e113e8975a399e40d7c0b1ddf4da6d3c59
│      version = v"0.5.7"
└    )
┌ Info: is_package_downloaded
└   sourcepath = "/home/fredrik/.julia/packages/PaddedViews/VweJw"
┌ Info: is_package_downloaded
│   project_file = "/home/fredrik/dev/pkg-dbg/MosaicViews/Project.toml"
│   pkg =
│    PackageSpec(
│      name = MosaicViews
│      uuid = e94cdb99-869f-56ef-bcf0-1ae2bcbe0389
│      version = v"0.2.4"
└    )
┌ Info: is_package_downloaded
└   sourcepath = nothing
ERROR: AssertionError: sourcepath !== nothing
```
where `MosaicViews` have neither a `git-tree-sha1` nor a `path` in its `PackageSpec`.

But there you are not running with `MosaicViews` as the current project.
```bash
$ cd $(mktemp -d)
$ git clone https://github.com/JuliaArrays/MosaicViews.jl . 2> /dev/null
$ pkg --project=. add ImageCore 2> /dev/null
$ pkg instantiate 
ERROR: LoadError: AssertionError: sourcepath !== nothing
```

--- a/src/Operations.jl
+++ b/src/Operations.jl
@@ -112,7 +112,12 @@ function is_instantiated(env::EnvCache)::Bool
     pkgs = load_all_deps(env)
     # If the top-level project is a package, ensure it is instantiated as well
     if env.pkg !== nothing
-        push!(pkgs, Types.PackageSpec(name=env.pkg.name, uuid=env.pkg.uuid, version=env.pkg.version, path=dirname(env.project_file)))
+        # Top-level project may already be in the manifest (cyclic deps)
+        # so only add it if it isn't there
+        idx = findfirst(x -> x.uuid == env.pkg.uuid, pkgs)
+        if idx === nothing
+            push!(pkgs, Types.PackageSpec(name=env.pkg.name, uuid=env.pkg.uuid, version=env.pkg.version, path=dirname(env.project_file)))
+        end
     end
     # Make sure all paths/artifacts exist
     return all(pkg -> is_package_downloaded(env.project_file, pkg), pkgs)

--- a/test/new.jl
+++ b/test/new.jl
@@ -2411,6 +2411,21 @@ end
             Pkg.add(path="A")
         end
     end
+    # test #2302
+    isolate(loaded_depot=true) do
+        cd_tempdir() do dir
+            Pkg.generate("A")
+            Pkg.generate("B")
+            git_init_and_commit("B")
+            Pkg.develop(path="B")
+            Pkg.activate("A")
+            Pkg.add(path="B")
+            git_init_and_commit("A")
+            Pkg.activate("B")
+            # This shouldn't error even though A has a dependency on B
+            Pkg.add(path="A")
+        end
+    end
 end
 
 @testset "Offline mode" begin

